components/collector/src/source_collectors/testng/tests.py

```diff
@@ -1,23 +1,46 @@
 """TestNG tests collector."""
 
 from typing import cast
+from xml.etree.ElementTree import Element  # nosec, Element is not available from defusedxml, but only used as type
 
 from base_collectors import XMLFileSourceCollector
 from collector_utilities.functions import parse_source_response_xml
-from model import SourceMeasurement, SourceResponses
+from model import Entities, Entity, SourceMeasurement, SourceResponses
 
 
 class TestNGTests(XMLFileSourceCollector):
     """Collector for TestNG tests."""
 
+    TEST_RESULT = dict(FAIL="failed", PASS="passed", SKIP="skipped")  # nosec, TestNG test status to test result mapping
+
     async def _parse_source_responses(self, responses: SourceResponses) -> SourceMeasurement:
         """Override to parse the tests for the TestNG XML."""
-        test_statuses_to_count = cast(list[str], self._parameter("test_result"))
+        test_results_to_count = [status.lower() for status in cast(list[str], self._parameter("test_result"))]
         test_count = 0
         total = 0
+        entities = Entities()
         for response in responses:
             tree = await parse_source_response_xml(response)
-            for test_status in test_statuses_to_count:
-                test_count += int(tree.get(test_status) or "0")
+            for test_result in test_results_to_count:
+                test_count += int(tree.get(test_result) or "0")
             total += int(tree.get("total") or "0")
-        return SourceMeasurement(value=str(test_count), total=str(total))
+            entities.extend(self.__entities(tree, test_results_to_count))
+        return SourceMeasurement(value=str(test_count), total=str(total), entities=entities)
+
+    @classmethod
+    def __entities(cls, tree: Element, test_results_to_count: list[str]) -> Entities:
+        """Transform the test methods into entities."""
+        entities = Entities()
+        for test_class in tree.findall(".//class"):
+            class_name = test_class.get("name", "")
+            for test_method in test_class.findall(".//test-method"):
+                test_result = cls.TEST_RESULT.get(test_method.get("status", ""))
```

Comment:
Is the default empty string required useful here? If I understand correctly, if it would fallback to the empty string, the TEST_RESULT.get() would raise a keyError anyway.

Reply:
Personally, in this case I would prefer the TEST_RESULT[test_method["status"]] syntax, or maybe even cleaner to define a status variable before

Reply:
I added "" as second argument to `Element.get()` because otherwise it returns None as default value which makes the type of the `Element.get()` result `Optional[str]` instead of `str`. But I'll see if I can refactor it so that it doesn't raise this question.